std/assembly/typedarray.ts

```diff
@@ -1718,3 +1762,59 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
   store<usize>(out, changetype<usize>(buffer) + <usize>byteOffset, offsetof<TArray>("dataStart"));
   return changetype<TArray>(out); // retains
 }
+
+// @ts-ignore: decorator
+@inline
+function SET<TArray extends ArrayBufferView, T, UArray extends ArrayBufferView, U>(target: TArray, source: UArray, offset: i32 = 0): void {
+  // need to assert at compile time that U is not a reference or a function
+  if (isReference<U>()) {
+    ERROR(E_NOTIMPLEMENTED);
+  }
+
+  // Uncaught RangeError: offset is out of bounds
+  if (offset < 0) throw new RangeError(E_INDEXOUTOFRANGE);
+  if (source.length + offset > target.length) throw new RangeError(E_INDEXOUTOFRANGE);
+
+  // if the types align and match, use memory.copy() instead of manual loop
+  if (isInteger<T>() == isInteger<U>() && alignof<T>() == alignof<U>()) {
+    memory.copy(
+      target.dataStart + (<usize>offset << alignof<T>()),
+      source.dataStart,
+      source.byteLength
+    );
+  } else {
+    let targetDataStart = target.dataStart + (<usize>offset << alignof<T>());
+    let sourceDataStart = source.dataStart;
+    let count = source.length;
+    for (let i = 0; i < count; i++) {
+      // if TArray is Uint8ClampedArray, then values must be clamped
+      if (target instanceof Uint8ClampedArray) {
+        if (isFloat<U>()) {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          store<T>(
+            targetDataStart + (<usize>i << alignof<T>()),
+            isFinite<U>(value)
+              // @ts-ignore: cast to T is valid for numeric types here
+              ? select<T>(0, select<T>(255, <T>value, value > 255), value < 0)
+              : 0
```

Comment:
```ts
<T>max<U>(0, min<U>(255, value))
```

Reply:
This optimization was a big deal! Thanks for the help.